Unity Build Cleanup

UnityAlex opened the pull request:
This change will remove the need for the following scrips:
* `.yamato/scripts/test_linux.sh`
* `.yamato/scripts/test_osx.sh`
* `.yamato/scripts/test_windows.cmd`
* `.yamato/scripts/build_linux.sh`
* `.yamato/scripts/build_osx.sh`
* `.yamato/scripts/build_windows.cmd`

They are replaced with:
* `.yamato/scripts/build_yamato.cmd`
* `.yamato/scripts/build_yamato.sh`

All of the build logic has been moved into `unity/CITools/BuildDriver` to have a single source of truth.

UnityAlex committed:
Initial commit for build work

mrvoorhe commented on .yamato/build_linux_x64_debug.yml:
```diff
@@ -11,7 +11,10 @@ variables:
 commands:
-  - ./.yamato/scripts/build_linux.sh Debug
+  - mkdir artifacts
+  - curl https://public-stevedore.unity3d.com/r/public/7za-linux-x64/e6c75fb7ffda_e6a295cdcae3f74d315361883cf53f75141be2e739c020035f414a449d4876af.zip --output artifacts/7za-linux-x64.zip
```
Now that I see how the zip downloading looks, maybe this would be better to move into C#.  Maybe put download & extraction behind `if(RunningOnYamato)`.  Or make it run locally but add a crude exists check caching scheme.

My concern is that sprinkling the steve hashes around is not ideal.

UnityAlex replied:
I'll have to take a look but the hashes are usually there in some form, just typically in the manifest file instead of being referenced directly like this.

mrvoorhe replied:
yes, I just meant that the has could be consolidated to one place.

UnityAlex replied:
Ah yeah I could rig up the usual bee + stevedore song and dance or just rig up a poor man's imitation since we only need it for this one thing.

mrvoorhe replied:
Poor man's is fine

UnityAlex replied:
Poor man's version has been implemented

UnityAlex committed:
Move 7zip download & extraction into build process

Use local 7z when not running on yamato

UnityAlex committed:
need to pass parameters correctly from yamato scripts

Need to pass special flag to ensure our args are only passed to the build script.

UnityAlex committed:
Running tests through CITools instead of separate scripts

Removing / updating TODO's after investigation.

mrvoorhe commented on unity/CITools/BuildDriver/CoreCLR.cs:
```diff
@@ -0,0 +1,84 @@
+        psi.Arguments = $"{gConfig.Architecture} {gConfig.Configuration}";
+        Utils.RunProcess(psi, gConfig);
+
+        // TODO: No PAL tests on windows?
```
Does this TODO need to be addressed?

UnityAlex replied:
Windows uses internal APIs and doesn't go through the PAL similar to mono.

mrvoorhe commented on unity/CITools/BuildDriver/CoreCLR.cs:
```diff
@@ -0,0 +1,84 @@
+        Utils.RunProcess(psi, gConfig);
+
+        string osString = RuntimeInformation.IsOSPlatform(OSPlatform.OSX) ? "OSX" : "Linux";
+        // TODO: We always run in Debug?
```
Does this TODO need to be addressed?  I ask because my routine is to write TODO's and then address them before landing, but sometimes I miss them so it's helpful if people call them out in my PR's.

UnityAlex replied:
I checked the coreclr documentation and this is how it's supposed to be run.

mrvoorhe commented on unity/CITools/BuildDriver/EmbeddingHost.cs:
```diff
@@ -0,0 +1,70 @@
+        // Run managed first
+        Build(gConfig, true);
+
+        // TODO: Why is this skipped?
```
TODO

mrvoorhe commented on unity/CITools/BuildDriver/EmbeddingHost.cs:
```diff
@@ -0,0 +1,70 @@
+
+        Utils.RunProcess(psi, gConfig);
+
+        // TODO: Check to see if we need to do a special dance on OSX arm similar to the nullgc
```
TODO

mrvoorhe commented on unity/CITools/BuildDriver/Program.cs:
```diff
@@ -0,0 +1,160 @@
+        Console.WriteLine("*****************************");
+
+        GlobalConfig gConfig = new (){ Architecture = architecture, Configuration = configuration, Silent = silent};
+        // TODO: Do we need to build this every time?
```
TODO

UnityAlex replied:
This was actually something I wanted to ask you. Is there anything in your changes that would make it so we don't need to build the `managed.sln` every time? I thought you had mentioned that you had added something that would kick off the build for it.

mrvoorhe replied:
That doesn't ring any bells.  It needs to be built every time.  It's the copying of the dlls to the coreclr artifact directory that I added logic for.  But that logic won't run here because coreclr may not have ran yet.

mrvoorhe commented on unity/CITools/BuildDriver/Program.cs:
```diff
@@ -0,0 +1,160 @@
+            NullGC.Build(gConfig);
+            CoreCLR.Build(gConfig);
+
+            // TODO: No longer needed with copying done in csproj?
```
TODO

UnityAlex replied:
This one you had a proposal to have this handled within the CSProj? I think the catch was that we'd have to build it twice or something to trigger the copy?

mrvoorhe replied:
Yes.  I'd land this first.  Then try replacing this with a rebuild of managed.sln in a follow on PR.
